Improve styles of filtering options for Clippy's lint list

CrazyRoka opened the pull request:
Partially solves #7958 

Updated styles for filtering options. It now uses dropdown menus.

![image](https://user-images.githubusercontent.com/19844144/144608479-cdd9de0b-f101-4d49-a135-0969efb01a11.png)

changelog: none

rust-highfive commented:
r? @camsteffen

(rust-highfive has picked a reviewer for you, use r? to override)

camsteffen commented:
The dropdowns look really nice! I think there should be some sort of always-visible indicator of what is selected. Possibly just the number of selected items, or maybe even just a different style to indicate that the default selection has been changed.

CrazyRoka commented:
I added badges to display selected items count. Have a look:

![image](https://user-images.githubusercontent.com/19844144/144719368-a2953657-2190-48c1-8536-58555957e5c8.png)

camsteffen commented:
Nice. I think the badge could be the same color as the lint category badges. Also can we see what it looks like on the right side of the button?

CrazyRoka commented:
I updated the code. Have a look:

- Badge color is now `#777`
- Badge is now located at the right side

![image](https://user-images.githubusercontent.com/19844144/144758211-1799ff95-558d-46d7-973b-a1ac70b80ac2.png)

xFrednet commented:
Also, a +1 from me for the dropdown menus, they look really nice :+1: 

The website also supports the different color themes from mdBook. The change theme button is in the upper-left corner. Can you make sure that your changes also support the change of color? You can find the available css values here: https://github.com/rust-lang/mdBook/blob/master/src/theme/css/variables.css

CrazyRoka commented:
@xFrednet @camsteffen Have a look on this video:

https://user-images.githubusercontent.com/19844144/144762393-14fd470f-9ce5-412c-aa03-569c96858d5b.mp4

xFrednet commented:
I really like it, nice work! :+1: 

Also cc: @flip1995 since you've suggested this tabbed filter layout. :upside_down_face:

camsteffen commented:
I am seeing JavaScript errors upon selecting options.

The button style seems to change from selected to deselected after clicking one of the options. Maybe this could be fixed with a CSS rule? If not it's probably fine.

CrazyRoka commented:
> I am seeing JavaScript errors upon selecting options.

I could see the error in Firefox browser, but not in Chrome. Replaced invalid code with correct one.

> The button style seems to change from selected to deselected after clicking one of the options. Maybe this could be fixed with a CSS rule? If not it's probably fine.

There was a style from bootstrap. I replaced it with custom one. Now selected drop-down should be slightly darker than others.

camsteffen commented:
I think it would be wise to use the Bootstrap dropdown js plugin ([link](https://bootstrapdocs.com/v3.3.6/docs/javascript/#dropdowns)), especially since we're already using Bootstrap CSS. This would be an added dependency though.

@rust-lang/clippy Everyone good with the general direction here?

flip1995 commented:
I like the general approach and how the drop down menus look. I don't know anything about JS / CSS, so I can't really review or give opinions on that front.

Only concern is, if and how it looks on mobile.

flip1995 commented on util/gh-pages/index.html:
```diff
@@ -25,7 +25,63 @@
+            background-color: var(--theme-popup-border);
+        }
+
+        .dropdown-menu .checkbox { 
```
There are some trailing whitespaces in this file.

CrazyRoka replied:
Removed all the trailing whitespaces

flip1995 reviewed (commented):
Selecting "All" has a small lag. This was especially noticeable on mobile.

Here's a link to a deployed version of this branch (rebased on current master): https://flip1995.github.io/rust-clippy/master/index.html 

One issue I found on mobile is, that the color scheme drop down is behind the other drop downs. Otherwise it LGTM on mobile as well.
![Screenshot_20211206-182635](https://user-images.githubusercontent.com/9744647/144893672-1f2b4605-779b-4f3f-8fa2-39d10fff9ae3.jpg)

camsteffen commented:
The "Filter" field could be on the same line and wrap responsively.

CrazyRoka commented:
> Selecting "All" has a small lag. This was especially noticeable on mobile.

I can see small lag on my modern laptop. Looks like selecting "All" will render all the lints at once and it causes Angular to take time building the DOM and parsing html:

![image](https://user-images.githubusercontent.com/19844144/144930133-16512009-5783-402d-aa51-ffffa9ccaf20.png)

We can add pagination to prevent this

CrazyRoka commented:
> The "Filter" field could be on the same line and wrap responsively.

Implemented this one.

> One issue I found on mobile is, that the color scheme drop down is behind the other drop downs.

Fixed this one

> I think it would be wise to use the Bootstrap dropdown js plugin (link), especially since we're already using Bootstrap CSS. This would be an added dependency though.

Does it support multi-select? As you can see, current dropdown supports selecting multiple entries. Also, it doesn't close dropdown, so user can select multiple items and close it by clicking on the button or somewhere in the page.

camsteffen commented:
Another thing we could do about the lag is make the lint panel expanded contents render lazily with something like `ngIf="expanded"`. That could be a standalone enhancement.

> Does it support multi-select? As you can see, current dropdown supports selecting multiple entries. Also, it doesn't close dropdown, so user can select multiple items and close it by clicking on the button or somewhere in the page.

I'm only judging by the docs, but I think the dropdown plugin is pretty unopinionated - you can put any HTML in the dropdown menu and you can close it programmatically like you currently are (but I would try out the default behavior first).

flip1995 commented:
Redeployed:  https://flip1995.github.io/rust-clippy/master/index.html

Another thing I noticed on mobile (tangentially related to this PR): The "Fork me on GitHub" banner in the top right corner slightly overlaps with the Lint groups dropdown. So when you try to open the dropdown by pressing on the error it redirects you to the Clippy GH repo, instead of opening the dropdown.

CrazyRoka commented:
> Another thing I noticed on mobile (tangentially related to this PR): The "Fork me on GitHub" banner in the top right corner slightly overlaps with the Lint groups dropdown. So when you try to open the dropdown by pressing on the error it redirects you to the Clippy GH repo, instead of opening the dropdown.

Fixed this issue with `clip-path` property. So, clickable area would look like this:

![image](https://user-images.githubusercontent.com/19844144/145070501-f2c78eeb-9b27-4cf3-8ad6-b7c1c2f1b065.png)

CrazyRoka commented:
> Another thing we could do about the lag is make the lint panel expanded contents render lazily with something like `ngIf="expanded"`. That could be a standalone enhancement.

Performance was improved. As you can see, Browser computation is now taking the most time after click:
![image](https://user-images.githubusercontent.com/19844144/145079690-ded8ddbc-7d6c-46b1-a9ec-d7144a62a69c.png)

CrazyRoka commented:
@flip1995 @camsteffen @xFrednet I went through all the reviews, and it looks like all the issues are resolved. Let me know if I missed something.
I haven't used the dropdown plugin, so. The reason is that I implemented this with raw `ng-click`, but this plugin will require me to rewrite it in `jquery` and add additional code to the rendered HTML.
@camsteffen If you strongly suggest using the plugin instead, I will add it.

camsteffen commented:
Oh I didn't realize it requires jQuery. That's fine then. I hate jQuery lol. I still need to do a full review but I feel confident about this PR overall.

xFrednet commented on util/gh-pages/index.html:
```diff
@@ -240,7 +307,7 @@
-<body>
+<body ng-app="clippy" ng-controller="lintList" ng-click="toggleDropdown(undefined, $event)">
     <div id="theme-icon" class="theme-icon">&#128396;</div>
     <ul id="theme-menu" class="theme-choice" style="display: none;">
```
The popup menus you created look a bit better than the current theme menu and they also close when the user interacts with the page body besides them. Would you mind updating the theme menu as well as part of this PR?

CrazyRoka replied:
I updated theme menu. Now it should close when interacting with the page

xFrednet commented:
The performance increase while searching is wonderful. I really like the changes!

flip1995 reviewed (approved):
This looks really good now. I love the perf improvement. Thank you for addressing all of our reviews so quickly!

Before this can be merged, please rebase your commits, in order to get rid of the merge commit.

CrazyRoka commented:
> This looks really good now. I love the perf improvement. Thank you for addressing all of our reviews so quickly!
> 
> Before this can be merged, please rebase your commits, in order to get rid of the merge commit.

I rebased my commits and added one new change:
- Theme menu now should be closed when interacting with the page
- Dropdowns are now implemented with directives (less code, more maintainability)
- Removed raw js handler for theme menu

camsteffen commented on util/gh-pages/index.html:
```diff
@@ -160,23 +229,28 @@
+            display: block;
+        }
+
+        .theme-choice>li {
```
nit: I'd prefer to keep the spacing around `>`

camsteffen commented on util/gh-pages/index.html:
```diff
@@ -25,7 +25,62 @@
+            color: var(--searchbar-fg);
+            border-color: var(--theme-popup-border);
+            filter: brightness(90%);
+        }
```
Bootstrap theming is normally done by downloading `bootstrap-theme.css` and modifying the values. We don't need an entire theme, but I think we should copy parts of that code, and add a comment saying as much.

CrazyRoka replied:
These variables are coming from rust mdBook styles. Which code do you mean?

camsteffen reviewed (commented):
How about aligning the text and checkboxes like this?

![image](https://user-images.githubusercontent.com/5565418/145917832-4bc3eb0d-a69a-484a-b0e0-81bfebfc8d51.png)

Feel free to clean up the commit history a bit as well.

CrazyRoka commented on util/gh-pages/index.html:
```diff
@@ -25,7 +25,62 @@
+
+         .dropdown-menu .checkbox:hover {
+            background-color: var(--theme-hover);
+        }
```
Fixed it

CrazyRoka commented on util/gh-pages/index.html:
```diff
@@ -25,7 +25,62 @@
+            font-weight: 400;
+            line-height: 1.42857143;
+            white-space: nowrap;
+            margin: 0;
```
Applied your suggestion

CrazyRoka commented:
@camsteffen I reviewed your suggestions and applied them. One comment is still under review, have a look. Let me know if you have any other concerns.

camsteffen commented:
@CrazyRoka just checking in. There's still one open thread. But honestly it's not super important and we can just wrap this up if you prefer.

xFrednet commented:
Hey @camsteffen, do we want to merge it, since @CrazyRoka hasn't replied yet? I like these changes and would like to have them merged before they conflict with new changes :upside_down_face:

camsteffen commented:
Okay. @bors r+

bors commented:
:pushpin: Commit 296af08b3c4ee2cd74b53e71279c8d2be254031e has been approved by `camsteffen`

<!-- @bors r=camsteffen 296af08b3c4ee2cd74b53e71279c8d2be254031e -->
<!-- homu: {"type":"Approved","sha":"296af08b3c4ee2cd74b53e71279c8d2be254031e","approver":"camsteffen"} -->

bors commented:
:hourglass: Testing commit 296af08b3c4ee2cd74b53e71279c8d2be254031e with merge b4cfdcdbc08e946b20ead368a01e6fe919b5514e...
<!-- homu: {"type":"BuildStarted","head_sha":"296af08b3c4ee2cd74b53e71279c8d2be254031e","merge_sha":"b4cfdcdbc08e946b20ead368a01e6fe919b5514e"} -->

bors commented:
:broken_heart: Test failed - [checks-action_test](https://github.com/rust-lang/rust-clippy/runs/5490211710?check_suite_focus=true)
<!-- homu: {"type":"BuildFailed","builder_url":"https://github.com/rust-lang/rust-clippy/runs/5490211710?check_suite_focus=true","builder_name":"checks-action_test"} -->

camsteffen commented:
@bors retry

bors commented:
:hourglass: Testing commit 296af08b3c4ee2cd74b53e71279c8d2be254031e with merge 8ae74da63b11ca298a30cc1af686019c7417938a...
<!-- homu: {"type":"BuildStarted","head_sha":"296af08b3c4ee2cd74b53e71279c8d2be254031e","merge_sha":"8ae74da63b11ca298a30cc1af686019c7417938a"} -->

bors commented:
:sunny: Test successful - [checks-action_dev_test](https://github.com/rust-lang/rust-clippy/runs/5490239015?check_suite_focus=true), [checks-action_remark_test](https://github.com/rust-lang/rust-clippy/runs/5490226997?check_suite_focus=true), [checks-action_test](https://github.com/rust-lang/rust-clippy/runs/5490347243?check_suite_focus=true)
Approved by: camsteffen
Pushing 8ae74da63b11ca298a30cc1af686019c7417938a to master...
<!-- homu: {"type":"BuildCompleted","approved_by":"camsteffen","base_ref":"master","builders":{"checks-action_test":"https://github.com/rust-lang/rust-clippy/runs/5490347243?check_suite_focus=true","checks-action_dev_test":"https://github.com/rust-lang/rust-clippy/runs/5490239015?check_suite_focus=true","checks-action_remark_test":"https://github.com/rust-lang/rust-clippy/runs/5490226997?check_suite_focus=true"},"merge_sha":"8ae74da63b11ca298a30cc1af686019c7417938a"} -->
